File: 101. Symmetric Tree.cpp
class Solution {
public:

    bool mirror(TreeNode* left, TreeNode* right) {

        // Both null
        if (left == nullptr && right == nullptr) {
            return true;
        }

        // One null
        if (left == nullptr || right == nullptr) {
            return false;
        }

        // Values differ
        if (left->val != right->val) {
            return false;
        }

        return mirror(left->left, right->right) &&
               mirror(left->right, right->left);
    }

    bool isSymmetric(TreeNode* root) {

        if (root == nullptr) {
            return true;
        }

        return mirror(root->left, root->right);
    }
};

File: 101. Symmetric Tree.py
class Solution:

    def mirror(self, left, right):

        # Both null
        if not left and not right:
            return True

        # One null
        if not left or not right:
            return False

        # Values differ
        if left.val != right.val:
            return False

        return (
            self.mirror(left.left, right.right) and
            self.mirror(left.right, right.left)
        )

    def isSymmetric(self, root):

        if not root:
            return True

        return self.mirror(root.left, root.right)
